### PR TITLE
csvparser refactor, prefix() and dialect() options

### PR DESCRIPTION
**Pull request by bazsi**
Hi,

This series of patches do a large scale overhaul of the csv-parser() parser, that I'd like to merge. 

A couple of notes:
 * the extraction of csv-scanner from csvparser seems like a big rewrite, but originally I made this as a series of 50+ smaller patches. Splitting it up would have been a lot of work, so at the end I decided not to do that. I can dig up my original (uncleaned) patch series if anyone is interested in the steps.
 * I've dropped one testcase from test_csvparser, which I've tried to explain in the commit msg
 * I've dropped the semi-generic g_string_list_find_first() function from the core, it was actually pretty csvparser specific, so I moved it there.
 * I've added a separate dialect() option replacing the escape-XXXX flags, but syslog-ng still supports
the old format and doesn't warn about the legacy format. We should decide which version this should land in (3.7.x or 3.8.x and add update warnings accordingly)
 * I've added a separate prefix() option, my original aim when starting the refactor, this is a new feature.

Please let me know if you want to separate patches into separate branches and let's decide what should go in which branches.

Thanks.

### CODE REVIEW

File: modules/csvparser/csvparser-grammar.ym [hunk 1]
```diff
@@ -102,15 +105,28 @@ parser_csv_delimiters_opts
...
+            $$ = flag | $2;
+            free($1);
+          }
         |					{ $$ = 0; }
         ;
 
+parser_csv_dialect
+        : string                                {
+                                                  gint mode = csv_parser_lookup_dialect($1);
+                                                  CHECK_ERROR(mode >= 0, @1, "unknown escape-mode() argument for csv-parser()");
```
Comment (ihrwein) [inline modules/csvparser/csvparser-grammar.ym]: `escape-mode` stayed here

File: modules/csvparser/csv-scanner.c [hunk 2]
```diff
@@ -0,0 +1,424 @@
...
+static const gboolean
+_match_string_delimiters_at_current_position(const char *input, GList *string_delimiters, int *result_length)
+{
+  GList *l;
+
+  for (l = string_delimiters; l; l = l->next)
+    {
+      gint len = strlen(l->data);
+
+      if (strncmp(input, l->data, strlen(l->data)) == 0)
```
Comment (ihrwein) [inline modules/csvparser/csv-scanner.c]: `len` can be used  here

File: modules/csvparser/csv-scanner.c [hunk 3]
```diff
@@ -0,0 +1,424 @@
...
+_skip_whitespace(const gchar **src)
+{
+  while (_is_whitespace_char(*src))
+    (*src)++;
+}
+
+static void
+_switch_to_next_column(CSVScanner *self)
+{
+  if (!self->current_column && self->src && self->src[0])
```
Comment (ihrwein) [inline modules/csvparser/csv-scanner.c]: Wouldn't it be better to move these two lines into `csv_scanner_parse_input()`? It's a kind of initialization before every `process()` call.

**Comment by ihrwein** [comment]
I checked the new parts and just got a glimpse of the old code. I placed my comments on the full change, but I have a longer comment about the grammar - csvparser interaction.

IMHO there should be a clear interface between the grammar and the csvparser logic. The `lookup_flag()` function violated this conceptual interface by moving some grammar logic into csvparser. For example, the grammar could define the KW_ESCAPE_NONE keyword, maybe it could even aggregate the flags into one flag.  I think the new lookup_dialect function does a similar thing, the escaping options could be defined as keywords and handled by the grammar, but you may have a good reason to do this which I missed.

What do you think?

File: modules/csvparser/csvparser-grammar.ym (hunk 1 above)
Comment (bazsi) [inline modules/csvparser/csvparser-grammar.ym]: fixed in 000623bf45deb8591a5379d7d693c5b0d8a65f1f

File: modules/csvparser/csv-scanner.c (hunk 2 above)
Comment (bazsi) [inline modules/csvparser/csv-scanner.c]: fixed in 17cc8858efb4e578db7bea499d94e660806f5239

File: modules/csvparser/csv-scanner.c (hunk 3 above)
Comment (bazsi) [inline modules/csvparser/csv-scanner.c]: The first call to csv_scanner_scan_next() must have current_column point to the first column and not the 2nd.

So we initialize the scanner when input() is called to have current_column == NULL, which either indicates that we have not even started processing the input yet (the first branch of the if), OR we are at the end, already having processed everything, the difference being whether unprocessed data is still available.

I have added a followup patch 5cebfb43d90e55d32c9bab11eb0bd368419e1a21, that should handle a potential NULL deref.

I am open to suggestions on improvements.

**Comment by bazsi** [comment]
It's not a good idea to make keywords from flags/other identifiers as those will be reserved in the entire grammar, not just in the flags() where one would use them.

In most cases, the top-level LogPipe that is being configured by the grammar has some code that directly supports the grammar. I agree that the coupling between the grammar and the underlying LogPipe is a bit too tight. There are numerous effects, like:

1) the self argument of setters is not of the proper types, as it would require logs of casts in the grammar
2) sometimes these setters manage memory in a way that makes the grammar simpler (for example they consume their arguments, so the grammar doesn't need to free them)
3) sometimes the grammar doesn't even use setters, but poke into the objects directly.

So as it is today, there are two responsibilities of LogPipe classes that are configured by the grammar:
1) they provide configurability to the grammar
2) they perform the actions (csv-parsing, rewrite rule, filters, whatever) that was configured

I can even agree that separating those two roles in thedesign would have been more fortunate, as we could make it easier to replace the grammar with something else. However, in the short term, I wouldn't change this, especially not one-by-one.

0) it's a lot of work, which if we stopped mid-way would just increase confusion
1) I think it is more important to improve tests, our unit tests are usually of the end-to-end kind, and libtest/testutils.h is ugly, a nicer test library would be badly needed
2) I also think that there are dark corners of the code (like csv-parser was before) which would be more important to do

Cheers,
Bazsi

File: modules/csvparser/csv-scanner.c (hunk 3 above)
Comment (ihrwein) [inline modules/csvparser/csv-scanner.c]: With this in mind It's completely reasonable, thanks for the explanation.

**Comment by ihrwein** [comment]
I did this coupling in one place in one of my projects and just never realized it before. Now I've to refactor that but syslog-ng is a much bigger project so it's fine to concentrate on tasks with higher priorities.

Maybe choosing a good test framework would be a good Hackathon project :)

Nevertheless, :+1:

**Comment by bazsi** [comment]
I would rebase the branch then, and merge in the latest additions. Any objections to the rebase?

**Comment by ihrwein** [comment]
I've no one.

**Comment by bazsi** [comment]
I have now squashed in the changed related to review notes, so the branch can be merged.

Please merge. Thanks.

**Comment by lbudai** [comment]
ok, then it will go to 3.8

**Pull request merged**